File: agentic-loops/agent_loop.py
"""
Production Agentic Loop — Python
pip install anthropic tenacity structlog
export ANTHROPIC_API_KEY=sk-ant-...
"""

import os
import json
import time
import logging
import structlog
from typing import Callable, Any
from anthropic import Anthropic, APIStatusError, APIConnectionError, RateLimitError
from tenacity import retry, stop_after_attempt, wait_exponential, retry_if_exception_type

# ── Logging ───────────────────────────────────────────────────────────────────
structlog.configure(
    processors=[
        structlog.processors.TimeStamper(fmt="iso"),
        structlog.processors.add_log_level,
        structlog.dev.ConsoleRenderer(),
    ]
)
log = structlog.get_logger()

# ── Client ────────────────────────────────────────────────────────────────────
client = Anthropic(
    api_key=os.environ["ANTHROPIC_API_KEY"],
    timeout=60.0,
    max_retries=0,  # we handle retries manually via tenacity
)

# ── Tool registry ─────────────────────────────────────────────────────────────
# Register callables here: name → python function
TOOL_REGISTRY: dict[str, Callable[..., Any]] = {}

def register_tool(name: str):
    def decorator(fn: Callable):
        TOOL_REGISTRY[name] = fn
        return fn
    return decorator

# ── Example tools (replace with your real ones) ───────────────────────────────
@register_tool("get_weather")
def get_weather(city: str) -> str:
    # TODO: call real weather API
    return f"72°F, sunny in {city}"

@register_tool("search_web")
def search_web(query: str) -> str:
    # TODO: call real search API
    return f"Top results for: {query}"

@register_tool("run_python")
def run_python(code: str) -> str:
    import subprocess, sys
    result = subprocess.run(
        [sys.executable, "-c", code],
        capture_output=True, text=True, timeout=10
    )
    return result.stdout or result.stderr

# ── Tool definitions for the API ──────────────────────────────────────────────
TOOLS = [
    {
        "name": "get_weather",
        "description": "Get current weather for a city.",
        "input_schema": {
            "type": "object",
            "properties": {"city": {"type": "string"}},
            "required": ["city"],
        },
    },
    {
        "name": "search_web",
        "description": "Search the web for a query.",
        "input_schema": {
            "type": "object",
            "properties": {"query": {"type": "string"}},
            "required": ["query"],
        },
    },
    {
        "name": "run_python",
        "description": "Execute Python code and return stdout/stderr.",
        "input_schema": {
            "type": "object",
            "properties": {"code": {"type": "string"}},
            "required": ["code"],
        },
    },
]

# ── Retryable API call ────────────────────────────────────────────────────────
@retry(
    retry=retry_if_exception_type((APIConnectionError, RateLimitError)),
    wait=wait_exponential(multiplier=1, min=2, max=30),
    stop=stop_after_attempt(5),
)
def call_api(messages: list, system: str | None = None) -> Any:
    kwargs = dict(
        model="claude-opus-4-6",
        max_tokens=4096,
        tools=TOOLS,
        messages=messages,
    )
    if system:
        kwargs["system"] = system
    return client.messages.create(**kwargs)

# ── Core agent loop ───────────────────────────────────────────────────────────
def run_agent(
    user_message: str,
    system: str | None = None,
    max_iterations: int = 20,
    on_tool_call: Callable[[str, dict, str], None] | None = None,
) -> str:
    """
    Run a full agentic loop.

    Args:
        user_message:   The task to complete.
        system:         Optional system prompt.
        max_iterations: Hard cap on loop turns.
        on_tool_call:   Optional callback(tool_name, tool_input, result).

    Returns:
        Final text response from the model.

    Raises:
        RuntimeError on unrecoverable API error or unknown tool.
    """
    messages = [{"role": "user", "content": user_message}]
    iteration = 0
    start_time = time.time()

    log.info("agent.start", message=user_message[:120])

    while iteration < max_iterations:
        iteration += 1
        log.info("agent.iteration", n=iteration)

        try:
            response = call_api(messages, system=system)
        except APIStatusError as e:
            log.error("agent.api_error", status=e.status_code, body=e.message)
            raise RuntimeError(f"API error {e.status_code}: {e.message}") from e

        messages.append({"role": "assistant", "content": response.content})

        log.info("agent.response", stop_reason=response.stop_reason,
                 usage=response.usage.model_dump())

        # ── Done ──────────────────────────────────────────────────────────────
        if response.stop_reason == "end_turn":
            elapsed = round(time.time() - start_time, 2)
            log.info("agent.done", iterations=iteration, elapsed_s=elapsed)
            for block in response.content:
                if hasattr(block, "text"):
                    return block.text
            return ""

        # ── Tool use ──────────────────────────────────────────────────────────
        if response.stop_reason == "tool_use":
            tool_results = []
            for block in response.content:
                if block.type != "tool_use":
                    continue

                tool_name = block.name
                tool_input = block.input

                if tool_name not in TOOL_REGISTRY:
                    err = f"Tool '{tool_name}' not registered."
                    log.error("agent.tool_missing", tool=tool_name)
                    tool_results.append({
                        "type": "tool_result",
                        "tool_use_id": block.id,
                        "content": err,
                        "is_error": True,
                    })
                    continue

                log.info("agent.tool_call", tool=tool_name, input=tool_input)
                try:
                    result = str(TOOL_REGISTRY[tool_name](**tool_input))
                except Exception as exc:
                    result = f"Tool error: {exc}"
                    log.error("agent.tool_error", tool=tool_name, error=str(exc))

                log.info("agent.tool_result", tool=tool_name,
                         result_len=len(result))

                if on_tool_call:
                    on_tool_call(tool_name, tool_input, result)

                tool_results.append({
                    "type": "tool_result",
                    "tool_use_id": block.id,
                    "content": result,
                })

            messages.append({"role": "user", "content": tool_results})
            continue

        # ── Unexpected stop reason ────────────────────────────────────────────
        log.warning("agent.unexpected_stop", stop_reason=response.stop_reason)
        break

    raise RuntimeError(f"Agent did not complete within {max_iterations} iterations.")


# ── Entry point ───────────────────────────────────────────────────────────────
if __name__ == "__main__":
    result = run_agent(
        user_message="Search the web for the latest news on AI agents, then write a 3-bullet summary.",
        system="You are a research assistant. Be concise.",
    )
    print("\n=== FINAL ANSWER ===")
    print(result)

File: go/src/agent_loop.go
// Plug-and-play Agentic Loop — Go
// Multi-provider via ModelRouter (Claude / DeepSeek / OpenAI).
// All tools are real implementations wired to the shared GoToolRegistry.
// No stubs. No hardcoded API clients.
//
// export MODEL_PROVIDER=anthropic|deepseek|openai
// export ANTHROPIC_API_KEY / DEEPSEEK_API_KEY / OPENAI_API_KEY
// go run ./src/... agent -task "your question"

package main

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strings"
)

// AgentLoop is a minimal single-shot agentic loop built on ModelRouter + GoToolRegistry.
// For multi-step ReAct behaviour, use RunReact() in react_loop.go instead.
func AgentLoop(ctx context.Context, userMessage string, maxIterations int, systemPrompt string) (string, error) {
	router, err := NewModelRouter("", "")
	if err != nil {
		return "", fmt.Errorf("agent_loop: model router init: %w", err)
	}

	messages := []map[string]interface{}{
		{"role": "user", "content": userMessage},
	}

	for i := 0; i < maxIterations; i++ {
		resp, err := router.Create(ctx, CreateOptions{
			Messages:  messages,
			Tools:     GoToolSchemas,
			System:    systemPrompt,
			MaxTokens: 4096,
		})
		if err != nil {
			return "", fmt.Errorf("agent_loop iteration %d: %w", i+1, err)
		}

		messages = append(messages, resp.ToAssistantMessage())

		if resp.StopReason == "end_turn" {
			for _, block := range resp.Content {
				if block.Type == BlockText && strings.TrimSpace(block.Text) != "" {
					return strings.TrimSpace(block.Text), nil
				}
			}
		}

		if resp.StopReason == "tool_use" {
			var toolResults []interface{}
			for _, block := range resp.Content {
				if block.Type != BlockToolUse {
					continue
				}
				if block.Name == "finish" {
					if answer, ok := block.Input["answer"].(string); ok {
						return answer, nil
					}
				}
				fn, ok := GoToolRegistry[block.Name]
				var result string
				isError := false
				if !ok {
					result = fmt.Sprintf("Tool '%s' is not registered", block.Name)
					isError = true
				} else {
					result, err = fn(block.Input)
					if err != nil {
						result = fmt.Sprintf("Tool error (%s): %v", block.Name, err)
						isError = true
					}
				}
				inp, _ := json.Marshal(block.Input)
				fmt.Fprintf(os.Stderr, "[agent] ⚡ %s(%s) → %s\n",
					block.Name, truncateStr(string(inp), 120), truncateStr(result, 120))
				toolResults = append(toolResults, map[string]interface{}{
					"type":        "tool_result",
					"tool_use_id": block.ID,
					"content":     result,
					"is_error":    isError,
				})
			}
			messages = append(messages, map[string]interface{}{
				"role":    "user",
				"content": toolResults,
			})
			continue
		}

		fmt.Fprintf(os.Stderr, "[agent] unexpected stop_reason=%s — stopping\n", resp.StopReason)
		break
	}

	return "", fmt.Errorf("AgentLoop: did not finish within %d iterations", maxIterations)
}

// AgentLoopDemo runs a demo task using the full production AgentLoop.
func AgentLoopDemo() {
	ctx := context.Background()
	task := "Use bash to find the current date and time, then report it."
	fmt.Fprintf(os.Stderr, "[agent_loop_demo] task: %s\n", task)
	answer, err := AgentLoop(ctx, task, 10, "You are a helpful assistant. Use the provided tools to complete tasks.")
	if err != nil {
		fmt.Fprintf(os.Stderr, "[agent_loop_demo] error: %v\n", err)
		os.Exit(1)
	}
	fmt.Println(answer)
}
